src/VisualStudio/Core/Def/TaskList/ExternalErrorDiagnosticUpdateSource.cs

```diff
@@ -139,385 +111,170 @@ public ImmutableArray<DiagnosticData> GetBuildErrors()
-                // for additional details.
-                _taskQueue.ScheduleTask(
-                    "OnDocumentChanged",
-                    () =>
-                    {
-                        using var argsBuilder = TemporaryArray<DiagnosticsUpdatedArgs>.Empty;
-                        AddArgsToClearBuildOnlyDocumentErrors(ref argsBuilder.AsRef(), e.OldSolution, e.ProjectId, e.DocumentId);
-                        ProcessAndRaiseDiagnosticsUpdated(argsBuilder.ToImmutableAndClear());
-                    },
-                    _disposalToken);
-                break;
-
-            case WorkspaceChangeKind.ProjectAdded:
-            case WorkspaceChangeKind.DocumentAdded:
-            case WorkspaceChangeKind.ProjectChanged:
-            case WorkspaceChangeKind.SolutionChanged:
-            case WorkspaceChangeKind.AdditionalDocumentAdded:
-            case WorkspaceChangeKind.AnalyzerConfigDocumentAdded:
-                break;
-
-            default:
-                throw ExceptionUtilities.UnexpectedValue(e.Kind);
-        }
+            await ClearPreviousAsync(projectId: projectId, cancellationToken).ConfigureAwait(false);
+        });
     }
 
+    /// <summary>
+    /// Called serially in response to the sln build UI context.
+    /// </summary>
     internal void OnSolutionBuildStarted()
     {
-        // Build just started, create the state and fire build in progress event.
         _ = GetOrCreateInProgressState();
-    }
-
-    internal void OnSolutionBuildCompleted()
-    {
-        // Building is done, so reset the state
-        // and get local copy of in-progress state.
-        var inProgressState = ClearInProgressState();
 
-        // Enqueue build/live sync in the queue.
-        _taskQueue.ScheduleTask("OnSolutionBuild", async () =>
+        _taskQueue.AddWork(async cancellationToken =>
         {
-            // nothing to do
-            if (inProgressState == null)
-            {
-                return;
-            }
-
-            // Mark the status as updated to refresh error list before we invoke 'SyncBuildErrorsAndReportAsync', which can take some time to complete.
-            OnBuildProgressChanged(inProgressState, BuildProgress.Updated);
-
-            // We are about to update live analyzer data using one from build.
-            // pause live analyzer
-            using var operation = _notificationService.Start("BuildDone");
-            if (_diagnosticService is DiagnosticAnalyzerService diagnosticService)
-                await SyncBuildErrorsAndReportOnBuildCompletedAsync(diagnosticService, inProgressState).ConfigureAwait(false);
-
-            // Mark build as complete.
-            OnBuildProgressChanged(inProgressState, BuildProgress.Done);
-        }, GetApplicableCancellationToken(inProgressState));
+            await ClearPreviousAsync(projectId: null, cancellationToken).ConfigureAwait(false);
```

Comment:
so i feel like the state object should be captuerd here.  otherwise, this lambda can run in the queue after other UI operations come in and reset/clear/recreate the 'in progress state'.  I don't mind initializing outside of the AddWork... but accessing within seems fundamentally racey.

Reply:
talked offline - state isn't used here